File: src/engine/adapters/mongo/caps.ts
import type { Caps } from '@shared/caps';

// §5.1's mongodb row: document-shaped, cursor pagination, estimate-only count, no FK
// navigation (§8.5), a shell-style console (§8.14). definition: true as of P19 D12 — a
// collection's creation options + validator, via mongo/definition.ts.
export const mongoCaps: Caps = {
  tabular: false,
  documents: true,
  keyValue: false,
  stream: false,
  defaultPageKind: 'document',
  sql: true,
  definition: true,
  describe: true,
  projection: true,
  serverFilter: true,
  exactCount: false,
  pagination: 'cursor',
  foreignKeys: false,
  // mutate.ts implements insert (insertOne), update (replaceOne) and delete (deleteOne).
  canInsert: true,
  canUpdate: true,
  canDelete: true,
  writable: true,
  transactions: false,
  cancel: true,
  fileTransfer: false,
};

File: apps/kira-studio/internal/adapters/mongo/caps.go
package mongo

import (
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/adapters"
	"github.com/kirathecat/kira-studio/apps/kira-studio/internal/page"
)

// caps is caps.ts's mongoCaps, literally (§4.2, C18). Document-shaped, cursor pagination,
// estimate-only count, no FK navigation, a shell-style console. Definition is true (P19 D12) —
// a collection's creation options + validator, via definition.go.
var caps = adapters.Caps{
	Tabular:         false,
	Documents:       true,
	KeyValue:        false,
	Stream:          false,
	KeyBrowser:      false,
	DefaultPageKind: page.PageKindDocument,
	SQL:             true,
	Definition:      true,
	Describe:        true,
	Projection:      true,
	ServerFilter:    true,
	ExactCount:      false,
	Pagination:      adapters.PaginationCursor,
	ForeignKeys:     false,
	// mutate.go implements insert (InsertOne), update (ReplaceOne) and delete (DeleteOne).
	CanInsert:    true,
	CanUpdate:    true,
	CanDelete:    true,
	Writable:     true,
	Transactions: false,
	Cancel:       true,
	FileTransfer: false,
}
